common/chains/client/multi_node_client.go

```diff
@@ -0,0 +1,310 @@
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+)
+
+var (
+	// PromMultiNodeClientRPCNodeStates reports current RPC node state
+	PromMultiNodeClientRPCNodeStates = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "client_rpc_node_states",
+		Help: "The number of RPC nodes currently in the given state for the given chain",
+	}, []string{"chainFamily", "chainId", "state"})
+)
+
+const (
+	NodeSelectionMode_HighestHead     = "HighestHead"
+	NodeSelectionMode_RoundRobin      = "RoundRobin"
+	NodeSelectionMode_TotalDifficulty = "TotalDifficulty"
+	NodeSelectionMode_PriorityLevel   = "PriorityLevel"
+)
+
+type NodeSelector[
+	CHAIN_ID types.ID,
+	BLOCK_HASH types.Hashable,
+	HEAD types.Head[BLOCK_HASH],
+	SUB types.Subscription,
+	RPC_CLIENT NodeClientAPI[CHAIN_ID, BLOCK_HASH, HEAD, SUB],
+] interface {
+	// Select returns a Node, or nil if none can be selected.
+	// Implementation must be thread-safe.
+	Select() Node[CHAIN_ID, BLOCK_HASH, HEAD, SUB, RPC_CLIENT]
+	// Name returns the strategy name, e.g. "HighestHead" or "RoundRobin"
+	Name() string
+}
+
+type MultiNodeClient[
+	CHAIN_ID types.ID,
+	BLOCK_HASH types.Hashable,
+	HEAD types.Head[BLOCK_HASH],
+	SUB types.Subscription,
+	RPC_CLIENT NodeClientAPI[CHAIN_ID, BLOCK_HASH, HEAD, SUB],
+] interface {
+	Dial(context.Context) error
+	Close() error
+	NodeStates() map[string]string
+	SelectNode() Node[CHAIN_ID, BLOCK_HASH, HEAD, SUB, RPC_CLIENT]
+
+	runLoop()
+	nLiveNodes() (int, int64, *utils.Big)
+	report()
+}
+
+func ContextWithDefaultTimeout() (ctx context.Context, cancel context.CancelFunc) {
+	return context.WithTimeout(context.Background(), queryTimeout)
+}
+
+type multiNodeClient[
+	CHAIN_ID types.ID,
+	BLOCK_HASH types.Hashable,
+	HEAD types.Head[BLOCK_HASH],
+	SUB types.Subscription,
+	RPC_CLIENT NodeClientAPI[CHAIN_ID, BLOCK_HASH, HEAD, SUB],
+] struct {
+	utils.StartStopOnce
+	nodes               []Node[CHAIN_ID, BLOCK_HASH, HEAD, SUB, RPC_CLIENT]
+	sendonlys           []Node[CHAIN_ID, BLOCK_HASH, HEAD, SUB, RPC_CLIENT]
+	chainID             CHAIN_ID
+	logger              logger.Logger
+	selectionMode       string
+	noNewHeadsThreshold time.Duration
+	nodeSelector        NodeSelector[CHAIN_ID, BLOCK_HASH, HEAD, SUB, RPC_CLIENT]
+	chainFamily         string
+
+	activeMu   sync.RWMutex
+	activeNode Node[CHAIN_ID, BLOCK_HASH, HEAD, SUB, RPC_CLIENT]
+
+	chStop utils.StopChan
+	wg     sync.WaitGroup
+}
+
+func NewMultiNodeClient[
+	CHAIN_ID types.ID,
+	BLOCK_HASH types.Hashable,
+	HEAD types.Head[BLOCK_HASH],
+	SUB types.Subscription,
+	RPC_CLIENT NodeClientAPI[CHAIN_ID, BLOCK_HASH, HEAD, SUB],
+](
+	logger logger.Logger,
+	selectionMode string,
+	noNewHeadsThreshold time.Duration,
+	nodes []Node[CHAIN_ID, BLOCK_HASH, HEAD, SUB, RPC_CLIENT],
+	sendonlys []Node[CHAIN_ID, BLOCK_HASH, HEAD, SUB, RPC_CLIENT],
```

Comment:
Why are we using the same type for both of these?
Don't send only nodes require a much simpler API?

If we need the same for both, then there might be a simpler way to express the generics, by capturing this whole `Node[...]` up top.